### ISSUE TITLE
Odd weewx shutdown

### ISSUE THREAD

**Issue by ktrue** [gjr80/weewx-saratoga#12]
weewx crashed this morning, and the error seems persistent:

```
May  5 06:44:44 TRP3WX weewx[29126] INFO weewx.engine: Main loop exiting. Shutting engine down.
May  5 06:44:44 TRP3WX weewx[29126] INFO weewx.engine: Shutting down StdReport thread
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__: Caught unrecoverable exception:
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****  'pressure'
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****  Traceback (most recent call last):
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/weewxd", line 157, in main
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      engine.run()
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/weewx/engine.py", line 210, in run
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      self.dispatchEvent(weewx.Event(weewx.NEW_LOOP_PACKET, packet=packet))
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/weewx/engine.py", line 245, in dispatchEvent
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      callback(event)
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/weewx/wxservices.py", line 98, in new_loop_packet
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      self.do_calculations(event.packet, self.loop_calc_dict)
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/weewx/wxservices.py", line 123, in do_calculations
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      new_value = weewx.xtypes.get_scalar(obs_type, data_dict, self.db_manager)
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/weewx/xtypes.py", line 73, in get_scalar
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      return xtype.get_scalar(obs_type, record, db_manager)
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/user/wsxtypes.py", line 53, in get_scalar
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      return getattr(self, 'calc_%s' % obs_type)(obs_type, record, db_manager)
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****    File "/usr/share/weewx/user/wsxtypes.py", line 66, in calc_wet_bulb
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****      if record['outTemp'] is not None and record['pressure'] is not None and record['outHumidity'] is not None:
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****  KeyError: 'pressure'
May  5 06:44:44 TRP3WX weewx[29126] CRITICAL __main__:     ****  Exiting.
```

Help??

**Comment by gjr80** [comment]
Looks like a bug in the logic for determining whether the pre-requisites exist for calculating wet bulb. Also affect a few other WeeWX-Saratoga XTypes. Should be fixed at commit 11e37b9dbba907438b29910a90799c19770cdec5.

The actual error is due to `pressure` not being in the loop packet/archive record concerned, it's absence is unusual for a Davis station and it's worth looking at why that might have happened. Provided at least one of the three pressure fields (`pressure`, `barometer` or `altitude`) exists WeeWX should calculate the other two. Davis stations typically provide `barometer`, though they can provide `pressure` if LOOP2 packets are being used. A default WeeWX install does not use LOOP2 packets. Calculation of `pressure` requires values for station altitude, barometer, outside humidity, current temperature and temperature 12 hours ago. All of these should exist, though the 12 hour temperature could be missing if the station was offline earlier (ie 12 hours earlier). Is there any chance the station was off line for a period, even for only five minutes or so, enough to cause an archive record to be missed/empty?

**Comment by ktrue** [comment]
It's possible the station was offline for a while.  I'm feeding it from a Davis VP1+ (ancient) and through VirtualVP, so no LOOP2 packets are even available.

I've installed 11e37b9 and it had updated the archives with missing records.  At the next archive period, it seemed to update everything ok.  I think the problem is fixed.. thanks!
